tests/foreman/virtwho/test_ui_virtwho.py

```diff
@@ -420,3 +427,155 @@ def test_positive_virtwho_configs_widget(session, form_data):
+        restart_virt_who_service()
+        try:
+            deploy_validation()
+        except Exception as Error:
+            assert str(Error) == 'Failed to start virt-who service'
+
+
+@tier2
+def test_positive_virt_who_roles_permissions(session, test_name, form_data):
+    """Verify the virt-who roles can TRULY work.
+
+    :id: 4bef3b31-89d2-4502-b971-2437b920a3ff
+
+    :expectedresults:
+        Permissions:
+        1. Virt-who Reporter Role granting minimal set of permissions for virt-who
+        to upload the report, it can be used if you configure virt-who manually
+        and want to use user that has locked down account.
+        2. Virt-who Viewer Role granting permissions to see all configurations
+        including their configuration scripts, which means viewers could still
+        deploy the virt-who instances for existing virt-who configurations.
+        3. Virt-who Manager Role granting all permissions to manage virt-who configurations,
+        user needs this role to create, delete or update configurations.
+
+    """
+    username = gen_string('alpha')
+    password = gen_string('alpha')
+    with session:
+        # Create a virt-who config plugin
+        virt_who_name = gen_string('alpha')
+        form_data['name'] = virt_who_name
+        session.virtwho_configure.create(form_data)
+        session.virtwho_configure.read(virt_who_name)
+        # Create an user
+        session.user.create({
+            'user.login': username,
+            'user.mail': 'test@test.com',
+            'user.auth': 'INTERNAL',
+            'user.password': password,
+            'user.confirm': password,
+        })
+        # Create a virt-who config file manually
+        virtwho_cleanup()
+        create_etc_d_file(form_data, username, password)
+        restart_virt_who_service()
+        try:
+            deploy_validation(False)
+        except Exception as Error:
+            assert str(Error) == 'Failed to start virt-who service'
+        # Check the permissioin of Virt-who Reporter
+        session.user.update(
+            username,
+            {'roles.resources.assigned': ['Virt-who Reporter']}
+        )
+        assert session.user.search(username)[0]['Username'] == username
+        user = session.user.read(username)
+        assert user['roles']['resources']['assigned'] == ['Virt-who Reporter']
+        restart_virt_who_service()
+        try:
+            deploy_validation(False)
+        except Exception:
+            raise VirtWhoError(
+                'Virt-who Reporter should have the permission to upload the report')
+        with Session(test_name, username, password) as newsession:
+            try:
+                newsession.virtwho_configure.read(virt_who_name)
+            except Exception:
+                pass
+            else:
+                raise VirtWhoError(
+                    'Virt-who Reporter should not have the permission to read config')
```

Comment:
You may use `assertRaises` to instead

Reply:
thanks for the idea, but seems airgun does not use those Unittest methods, I will try to find some pythonic and simple way to instead.